java-checks/src/main/java/org/sonar/java/checks/StaticFieldInitializationCheck.java

```diff
@@ -46,55 +50,92 @@
   }
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
+    classWithSynchronizedMethod.push(false);
     withinStaticInitializer.push(false);
     methodUsesLocks.push(false);
     super.scanFile(context);
     withinStaticInitializer.clear();
     methodUsesLocks.clear();
+    classWithSynchronizedMethod.clear();
   }
 
   @Override
   public void visitNode(Tree tree) {
-    if (tree.is(Tree.Kind.METHOD)) {
-      methodUsesLocks.push(false);
-    }
-    if (tree.is(Tree.Kind.METHOD_INVOCATION)
-      && (locks.anyMatch((MethodInvocationTree) tree))
-      && methodUsesLocks.size() != 1) {
-      methodUsesLocks.pop();
-      methodUsesLocks.push(true);
-    }
-    if (hasSemantic() && tree.is(Tree.Kind.ASSIGNMENT)) {
-      AssignmentExpressionTree aet = (AssignmentExpressionTree) tree;
-      if (aet.variable().is(Tree.Kind.IDENTIFIER) && !isInSyncBlock() && !withinStaticInitializer.peek() && !isUsingLock()) {
-        IdentifierTree variable = (IdentifierTree) aet.variable();
-        if (isStaticNotVolatileObject(variable)) {
-          reportIssue(variable, "Synchronize this lazy initialization of '" + variable.name() + "'");
+    switch (tree.kind()) {
+      case CLASS:
+        classWithSynchronizedMethod.push(hasSynchronizedMethod((ClassTree) tree));
+        break;
+      case STATIC_INITIALIZER:
+        withinStaticInitializer.push(true);
+        break;
+      case METHOD:
+        methodUsesLocks.push(false);
+        break;
+      case METHOD_INVOCATION:
+        if (locks.anyMatch((MethodInvocationTree) tree) && methodUsesLocks.size() != 1) {
+          methodUsesLocks.pop();
+          methodUsesLocks.push(true);
         }
-      }
-    }
-    if (tree.is(Tree.Kind.STATIC_INITIALIZER)) {
-      withinStaticInitializer.push(true);
+        break;
+      case ASSIGNMENT:
+        AssignmentExpressionTree aet = (AssignmentExpressionTree) tree;
+        if (hasSemantic()
+          && aet.variable().is(Tree.Kind.IDENTIFIER)
+          && !isInSyncBlock()
+          && !isInStaticInitializer()
+          && !isUsingLock()
+          && isInClassWithSynchronizedMethod()) {
+          IdentifierTree variable = (IdentifierTree) aet.variable();
+          if (isStaticNotVolatileObject(variable)) {
+            reportIssue(variable, "Synchronize this lazy initialization of '" + variable.name() + "'");
+          }
+        }
+        break;
+      default:
+        // Do nothing
     }
     super.visitNode(tree);
   }
 
+  private boolean isInStaticInitializer() {
+    return withinStaticInitializer.peek();
+  }
+
+  private static Boolean hasSynchronizedMethod(ClassTree tree) {
+    return tree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .map(MethodTree::modifiers)
+      .anyMatch(modifiers -> ModifiersUtils.hasModifier(modifiers, Modifier.SYNCHRONIZED));
```

Comment:
flatMap(MethodTree::modifiers).anyMatch(Modifier.SYNCHRONIZED::equals) ? avoid the call to ModifiersUtils.

Reply:
`Modifier.SYNCHRONIZED` is not a tree, contrarily to the modifiers from the list. the steps to get there make it less clear from my point of view.